src/input_utils.py

```diff
@@ -0,0 +1,87 @@
+import attr
+import tensorflow as tf
+import tensorflow_text as tf_text
+
+from typing import List
+
+from google_research.etcmodel.models.modeling import EtcConfig 
+
+
+@attr.s
+class PretrainInputConfig(object):
+  """Config options for pretraining model input."""
+
+  image_size = attr.ib(default=224)
+  text_keys = attr.ib(factory=List)
+  patch_size = attr.ib(default=16)
+  patch_order = attr.ib(default='snake')
+  mlm_use_whole_word = attr.ib()  # type: bool
+
+
+def get_pretrain_example_decode_fn(tokenizer: tf_text.BertTokenizer,
+                                   input_config: PretrainInputConfig,
+                                   model_config: EtcConfig,
+                                   is_training: bool):
+  """Returns a decode function to parse a single example into Tensors."""
+
+  image_size = input_config.image_size
+  patch_size = input_config.patch_size
+  num_patch_per_row = image_size // patch_size
+
+  name_to_features = {'image_data': tf.io.FixedLenFeature([], tf.string)}
+  for k in input_config.text_keys:
+    name_to_features[k] = tf.io.FixedLenFeature([], tf.string, default_value='')
+
+
+  def patch_reorder(im, mode='snake'):
```

Comment:
Nit: "raster_scan" may be more specific here.  "snake" could mean a different pattern where we alternate every row between left-to-right and right-to-left.

Reply:
Nit: It could help to add a docstring mentioning the expected shape of the `im` input tensor and resulting output shape.

Reply:
Agreed, "snake" is typically a synonym of "zig-zag", which alternates left-right.